test/distributed/test_autograd.py

```diff
@@ -0,0 +1,247 @@
+    os.environ["MASTER_PORT"] = str(12355)
+
+    DistributedManager.initialize()
+
+    manager = DistributedManager()
+    assert manager.is_initialized()
+
+    tensor_dim = 4
+    tensor = (rank + 1) * torch.ones(
+        (rank + 2, tensor_dim), device=f"cuda:{rank}", dtype=torch.float32
+    )
+    tensor.requires_grad_(True)
+    sizes = [r + 2 for r in range(world_size)]
+
+    gathered_tensor = all_gather_v(tensor, sizes, dim=0, group=None)
+
+    expected_tensor = (
+        torch.arange(world_size, device=f"cuda:{rank}", dtype=torch.float32) + 1
+    )
+    expected_tensor = expected_tensor.view(-1, 1).expand(-1, tensor_dim).contiguous()
+    expected_tensor = expected_tensor.repeat_interleave(
+        repeats=torch.tensor(sizes, device=f"cuda:{rank}"), dim=0
+    )
+
+    assert torch.allclose(expected_tensor, gathered_tensor)
+
+    grad_out = torch.ones_like(gathered_tensor) * (-1)
+    gathered_tensor.backward(gradient=grad_out)
+
+    expected_grad = torch.ones_like(tensor) * (-1) * world_size
+    assert torch.allclose(tensor.grad, expected_grad)
+
+    del os.environ["RANK"]
+    del os.environ["WORLD_SIZE"]
+    del os.environ["MASTER_ADDR"]
+    del os.environ["MASTER_PORT"]
+
+    DistributedManager.cleanup()
+
+
+def run_test_indexed_all_gather_v(rank, world_size):
+    os.environ["RANK"] = f"{rank}"
+    os.environ["WORLD_SIZE"] = f"{world_size}"
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = str(12355)
+
+    DistributedManager.initialize()
+
+    manager = DistributedManager()
+    assert manager.is_initialized()
+
+    # this test case is not ideal as it quite similar to the non-indexed case
+    # however, it is a first start to test correctness in general
+    tensor_dim = 4
+    tensor = torch.arange(1, world_size + 1, device=f"cuda:{rank}", dtype=torch.float32)
+    tensor = tensor.view(-1, 1).expand(-1, tensor_dim).contiguous()
+    tensor = tensor.repeat_interleave(repeats=rank + 1, dim=0)
+    tensor.requires_grad_(True)
+
+    send_sizes = [rank + 1 for _ in range(world_size)]
+    recv_sizes = [r + 1 for r in range(world_size)]
+
+    scatter_indices = [
+        torch.nonzero(tensor[:, 0] == (r + 1)).view(-1) for r in range(world_size)
+    ]
+
+    gathered_tensor = indexed_all_gather(
+        tensor, scatter_indices, recv_sizes, send_sizes, group=None
+    )
+
+    expected_tensor = torch.ones(
+        (sum(recv_sizes), tensor_dim), device=f"cuda:{rank}", dtype=torch.float32
+    ) * (rank + 1)
+
+    assert torch.allclose(expected_tensor, gathered_tensor)
+
+    grad_out = torch.ones_like(gathered_tensor) * (-1)
+    gathered_tensor.backward(gradient=grad_out)
+
+    expected_grad = torch.ones_like(tensor) * (-1)
+    assert torch.allclose(tensor.grad, expected_grad)
+
+    del os.environ["RANK"]
+    del os.environ["WORLD_SIZE"]
+    del os.environ["MASTER_ADDR"]
+    del os.environ["MASTER_PORT"]
+
+    DistributedManager.cleanup()
```

Comment:
Do you think it is sufficient to have this one cleanup method or should we also have a per group cleanup like you had earlier?

Reply:
I needed my "per group" cleanup particularly because of my usage of things like "group id" etc.. I see two orthogonal things here actually. 

One thing would be to take care of the "set" environment variables in DistributedManager - which now requires these "del" statements in a lot of tests.

The other thing would be indeed a per group cleanup, i.e. basically a removal of all the things w.r.t. a certain process group - potentially including all its sub/orthogonal groups without having to reset the whole state.